servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,335 @@
+        super(name);
+        this.encoderSupplier = encoderSupplier;
+        this.decoderSupplier = decoderSupplier;
+    }
+
+    @Override
+    public Buffer encode(final Buffer src, final int offset, final int length, final BufferAllocator allocator) {
+        requireNonNull(src);
+        final int availableBytes = src.readableBytes() - offset;
+        if (offset < 0) {
+            throw new IllegalArgumentException("Invalid offset: " + offset + " (expected >= 0)");
+        }
+
+        if (length <= 0) {
+            throw new IllegalArgumentException("Invalid length: " + length + " (expected > 0");
+        }
+
+        if (length > availableBytes) {
+            throw new IllegalStateException("Invalid length: " + length + " (expected <= " + availableBytes);
+        }
+
+        final MessageToByteEncoder<ByteBuf> encoder = encoderSupplier.get();
+        final EmbeddedChannel channel = newEmbeddedChannel(encoder, allocator);
+
+        try {
+            ByteBuf origin = toByteBuf(src, offset, length);
+            channel.writeOutbound(origin);
+
+            // May produce footer
+            preparePendingData(channel);
+            return requireNonNull(drainChannelOutToSingleBuffer(channel, allocator));
+        } catch (Throwable e) {
+            LOGGER.error("Error while encoding with {}", name(), e);
+            safeCleanup(channel);
+            throwException(e);
+            return EMPTY_BUFFER;
+        }
+    }
+
+    @Override
+    public Publisher<Buffer> encode(final Publisher<Buffer> from, final BufferAllocator allocator) {
+        requireNonNull(from);
+        return from
+                .concat(succeeded(END_OF_STREAM))
+                .liftSync(subscriber -> new PublisherSource.Subscriber<Buffer>() {
+
+                    private final MessageToByteEncoder<ByteBuf> encoder = encoderSupplier.get();
+                    private final EmbeddedChannel channel = newEmbeddedChannel(encoder, allocator);
+
+                    @Override
+                    public void onSubscribe(PublisherSource.Subscription subscription) {
+                        subscriber.onSubscribe(subscription);
+                    }
+
+                    @Override
+                    public void onNext(Buffer next) {
+                        if (next == null) {
+                            return;
+                        }
+
+                        // onNext will produce AT-MOST N items (from upstream)
+                        // +1 for the encoding footer (ie. END_OF_STREAM)
+                        if (next == END_OF_STREAM) {
+                            // May produce footer
+                            preparePendingData(channel);
+                            Buffer buffer = drainChannelOutToSingleBuffer(channel, allocator);
+                            if (buffer != null) {
+                                subscriber.onNext(buffer);
+                            }
```

Comment:
Some encoders (like Bzip2) use blocks and do not produce any output until the block is full. We should request more data in this case.